apollo-federation/src/correctness/query_plan_soundness.rs

```diff
@@ -0,0 +1,660 @@
+                    let sub_selections =
+                        convert_requires_selections(schema, &sub_ty, &field.selections)?;
+                    converted.with_selections(sub_selections)
+                };
+                result.push(converted.into());
+            }
+            requires_selection::Selection::InlineFragment(inline) => {
+                let converted = if let Some(type_condition) = &inline.type_condition {
+                    executable::InlineFragment::with_type_condition(type_condition.clone())
+                } else {
+                    executable::InlineFragment::without_type_condition(ty.clone())
+                };
+                let sub_ty = converted.selection_set.ty.clone();
+                let sub_selections =
+                    convert_requires_selections(schema, &sub_ty, &inline.selections)?;
+                result.push(converted.with_selections(sub_selections).into());
+            }
+        }
+    }
+    Ok(result)
+}
+
+/// Get the `ast::FieldDefinition` for the given type and field name from the schema.
+fn get_field_definition(
+    schema: &ValidFederationSchema,
+    parent_type: &Name,
+    field_name: &Name,
+) -> Result<Node<ast::FieldDefinition>, FederationError> {
+    let parent_type_pos: CompositeTypeDefinitionPosition =
+        schema.get_type(parent_type.clone())?.try_into()?;
+    let field_def_pos = parent_type_pos.field(field_name.clone())?;
+    field_def_pos
+        .get(schema.schema())
+        .map(|component| component.node.clone())
+        .map_err(|err| err.into())
+}
+
+/// Collect `@requires` conditions from the fields used in the fetch response shape based on their
+/// definitions in the subgraph schema.
+/// - `response_shape`: the fetch response shape for the entity
+/// - Only consider the fields at the root level of the response shape.
+/// - The fields may have Boolean conditions. Then, the computed conditions will inherit them.
+/// - The collected conditions are merged into a single response shape.
+fn collect_require_condition(
+    supergraph_schema: &ValidFederationSchema,
+    subgraph_schema: &ValidFederationSchema,
+    federation_spec_definition: &FederationSpecDefinition,
+    response_shape: &ResponseShape,
+) -> Result<ResponseShape, FederationError> {
+    let requires_directive_definition =
+        federation_spec_definition.requires_directive_definition(subgraph_schema)?;
+    let parent_type = response_shape.default_type_condition();
+    let mut result = ResponseShape::new(parent_type.clone());
+    for (_key, defs) in response_shape.iter() {
+        for (_type_cond, defs_per_type_cond) in defs.iter() {
+            for variant in defs_per_type_cond.conditional_variants() {
+                let field_def = &variant.representative_field().definition;
+                for directive in field_def
+                    .directives
+                    .get_all(&requires_directive_definition.name)
+                {
```

Comment:
This deep nesting looks like it can be largely flattened:
```rust
for variant in response_shape.iter()
                             .flat_map(|(_, defs)| defs.iter())
                             .flat_map(|(_, per_type_cond)| per_type_cond.conditional_variants())
```